src/DocBlock/TypeExpression.php

```diff
@@ -285,54 +296,46 @@ public function allowsNull(): bool
 
     private function parse(): void
     {
-        $value = $this->value;
-
-        Preg::match(
-            '{^'.self::REGEX_TYPES.'$}x',
-            $value,
-            $matches
-        );
-
-        if ([] === $matches) {
-            return;
-        }
-
-        $this->typesGlue = $matches['glue'] ?? $this->typesGlue;
-
-        $index = '' !== $matches['nullable'] ? 1 : 0;
-
-        if ($matches['type'] !== $matches['types']) {
-            $this->isUnionType = true;
+        $index = 0;
+        while (true) {
+            Preg::match(
+                '{\G'.self::REGEX_TYPE.'(?:\h*(?<glue>[|&])\h*|$)}x',
+                $this->value,
+                $matches,
+                0,
+                $index
+            );
 
-            while (true) {
-                $innerType = $matches['type'];
+            if ([] === $matches) { // invalid phpdoc type
+                // TODO once all phpdoc types are parsed strictly using self::REGEX_TYPES,
+                // the parse cannot fail and we can throw here safely
+                return;
+            }
 
-                $newValue = Preg::replace(
-                    '/^'.preg_quote($innerType, '/').'(\h*[|&]\h*)?/',
-                    '',
-                    $value
-                );
+            if (!$this->isUnionType) {
+                if (($matches['glue'] ?? '') === '') {
+                    break;
+                }
 
-                $this->innerTypeExpressions[] = [
-                    'start_index' => $index,
-                    'expression' => $this->inner($innerType),
-                ];
+                $this->isUnionType = true;
```

Comment:
Can you explain why it's like that? `glue` can be either `|` or `&`, so why `$this->isUnionType` is set to `true` when `glue` is not empty?

Reply:
I belive it is there for a historical reasons, this PR does not change it, It just tells there are more types glued together, either with union or intersect.

Reply:
But previously this assignment was hidden behind different condition, so it looked cleaner 😉. This feels somehow weird.

Reply:
Previously the input was parsed twice, now it is parsed only one and the condition is correct as simple, when there is no glue, expr is parsed, when there is a glue, inner expressions are parsed. I do no think the impl. can be done better, if you think so, please advise.

Reply:
Nahh, it's good enough, I just wanted to discuss it to understand it better.